File: week_8_and_9/3_JavaScript/gradebook/my_solution.js
/*
U3.W8-9: Gradebook from Names and Scores

You will work with the following two variables.  The first, students, holds the names of four students.  
The second, scores, holds groups of test scores.  The relative positions of elements within the two 
variables match (i.e., 'Joseph' is the first element in students; his scores are the first value in scores.).

Do not alter the students and scores code.

I worked on this challenge by myself.

*/





// __________________________________________
// Write your code below.

// var students = ["Joseph", "Susan", "William", "Elizabeth"]

// var scores = [ [80, 70, 70, 100],
//                [85, 80, 90, 90],
//                [75, 70, 80, 75],
//                [100, 90, 95, 85] ]






// // __________________________________________
// // Write your code below.

// var average = function(array){
    
//    var  sum = 0
//     for (i=0;i<array.length;i++)
//     {
        
//         sum += array[i]
        
//     }
    
//     avg = sum/array.length
//     return avg
// }

// var gradebook = {}
// gradebook.Joseph = {}
// gradebook.Susan = {}
// gradebook.Susan.name = "Susan"
// gradebook.William = {}
// gradebook.Elizabeth = {}
// gradebook.William.testScores = scores[2]
// gradebook.Joseph.testScores = scores[0]
// gradebook.Susan.testScores = scores[1]
// gradebook.Elizabeth.testScores = scores[3]
// gradebook.addScore = function(name,score)
// {
 
 
// for(i=0;i<students.length;i++)
// {
//     if (name===students[i])
//     {
        
//     scores[i].push(score)
//     }
    
    
// }

    
// }

// gradebook.getAverage = function(name)
// {
//  //a = Object.getOwnPropertyNames(gradebook)
 
// for(i=0;i<students.length;i++)
// {
//     if (name===students[i])
//     {
        
//     return average(scores[i])
//     }
    
    
// }

    
// }






// __________________________________________
// Refactored Solution

// Couldnt really think of anyway to refactor this code.

var students = ["Joseph", "Susan", "William", "Elizabeth"]

var scores = [ [80, 70, 70, 100],
               [85, 80, 90, 90],
               [75, 70, 80, 75],
               [100, 90, 95, 85] ]






// __________________________________________
// Write your code below.

var average = function(array){
    
   var  sum = 0
    for (i=0;i<array.length;i++)
    {
        
        sum += array[i]
        
    }
    
    avg = sum/array.length
    return avg
}

var gradebook = {}
gradebook.Joseph = {}
gradebook.Susan = {}
gradebook.Susan.name = "Susan"
gradebook.William = {}
gradebook.Elizabeth = {}
gradebook.William.testScores = scores[2]
gradebook.Joseph.testScores = scores[0]
gradebook.Susan.testScores = scores[1]
gradebook.Elizabeth.testScores = scores[3]
gradebook.addScore = function(name,score)
{
 
 
for(i=0;i<students.length;i++)
{
    if (name===students[i])
    {
        
    scores[i].push(score)
    }
    
    
}

    
}

gradebook.getAverage = function(name)
{
 
 
for(i=0;i<students.length;i++)
{
    if (name===students[i])
    {
        
    return average(scores[i])
    }
    
    
}

    
}



// __________________________________________
// Driver Code:  Do not alter code below this line.


function assert(test, message, test_number) {
  if (!test) {
    console.log(test_number + "false");
    throw "ERROR: " + message;
  }
  console.log(test_number + "true");
  return true;
}

assert(
  (average instanceof Function),
  "The value of average should be a Function.\n",
  "1. "
)

assert(
  average([1, 2, 3]) === 2,
  "average should return the average of the elements in the array argument.\n",
  "2. "
)

assert(
  (gradebook instanceof Object),
  "The value of gradebook should be an Object.\n",
  "3. "
)

assert(
  (gradebook["Elizabeth"] instanceof Object),
  "gradebook's Elizabeth property should be an object.",
  "4. "
)

assert(
  (gradebook.William.testScores === scores[2]),
  "William's testScores should equal the third element in scores.",
  "5. "
)

assert(
  (gradebook.addScore instanceof Function),
  "The value of gradebook's addScore property should be a Function.",
  "6. "
)

gradebook.addScore("Susan", 80)

assert(
  (gradebook.Susan.testScores.length === 5
   && gradebook.Susan.testScores[4] === 80),
  "Susan's testScores should have a new score of 80 added to the end.",
  "7. "
)

assert(
  (gradebook.getAverage instanceof Function),
  "The value of gradebook's getAverage property should be a Function.",
  "8. "
)

assert(
  (gradebook.getAverage("Joseph") === 80),
  "gradebook's getAverage should return 80 if passed 'Jospeh'.",
  "9. "
)


//Reflections

// What parts of your strategy worked? What problems did you face?

// I was out of touch with javascript and had to read up old assignments to get myself familiar with the syntax ,creating objects etc.

// What questions did you have while coding? What resources did you find to help you answer them?

// Yes I hadnt done any javascript coding for a while so had to read up before I could start the assignment.

// What concepts are you having trouble with, or did you just figure something out? If so, what?

// None

// Did you learn any new skills or tricks?

// No

// How confident are you with each of the Learning Competencies?

// I need some more javascript practise.

// Which parts of the challenge did you enjoy?

// Getting the driver code to work is always enjoyable

// Which parts of the challenge did you find tedious?

// None.

File: week_7/8_SQLite_gem/my_solution.rb
# U3.W7: BONUS Using the SQLite Gem

# I worked on this challenge by myself

require 'sqlite3'


$db = SQLite3::Database.open "congress_poll_results.db"


def print_arizona_reps
  puts "AZ REPRESENTATIVES"
  az_reps = $db.execute("SELECT name FROM congress_members WHERE location = 'AZ'")
  az_reps.each { |rep| puts rep }
end

def print_longest_serving_reps(minimum_years)  #sorry guys, oracle needs me, i didn't finish this!
  puts "LONGEST SERVING REPRESENTATIVES"
  long_reps =  $db.execute("SELECT name FROM congress_members WHERE years_in_congress > #{minimum_years}")
 

  long_reps.each do |rep|

     year = $db.execute("SELECT years_in_congress FROM congress_members WHERE name = '#{rep[0]}'")

     puts "#{rep[0]} - #{year[0][0]}"
    
 end
end

def print_lowest_grade_level_speakers(grade)
  puts "LOWEST GRADE LEVEL SPEAKERS (less than < 8th grade)"
  low_grade = $db.execute("SELECT name from congress_members WHERE grade_current < #{grade}")
  low_grade.each {|rep| puts rep}
end


def print_state_rep(*states)
puts "State REPRESENTATIVES"
puts

states.each do |x|

state_rep = $db.execute("SELECT name from congress_members WHERE location = '#{x}'")
 
 puts "#{x} REPRESENTATIVES"
 puts 
 state_rep.each {|rep| puts "#{rep[0]}"}
 puts

end
	end

def print_separator
  puts 
  puts "------------------------------------------------------------------------------"
  puts 
end


print_arizona_reps

print_separator

print_longest_serving_reps(35)
# TODO - Print out the number of years served as well as the name of the longest running reps
# output should look like:  Rep. C. W. Bill Young - 41 years

print_separator
print_lowest_grade_level_speakers(8)
# TODO - Need to be able to pass the grade level as an argument, look in schema for "grade_current" column

# TODO - Make a method to print the following states representatives as well:
# (New Jersey, New York, Maine, Florida, and Alaska)
print_separator
print_state_rep("NJ","NY","ME","FL","AK")


##### BONUS #######
# TODO (bonus) - Stop SQL injection attacks!  Statmaster learned that interpolation of variables in SQL statements leaves some security vulnerabilities.  Use the google to figure out how to protect from this type of attack.

# TODO (bonus)
# Create a listing of all of the Politicians and the number of votes they recieved
# output should look like:  Sen. John McCain - 7,323 votes (This is an example, yours will not return this value, it should just 
#    have a similar format)
# Create a listing of each Politician and the voter that voted for them
# output should include the senators name, then a long list of voters separated by a comma
#
# * you'll need to do some join statements to complete these last queries!


# REFLECTION- Include your reflection as a comment below.

# How does the sqlite3 gem work?  What is the variable `$db` holding?  

# $db holds the Database congress_poll_results. 

# Try to use your knowledge of ruby and OO to decipher this as well as h
# ow the `#execute` method works.  Take a stab at explaining the line 
# `$db.execute("SELECT name FROM congress_members WHERE years_in_congress 
#   > #{minimum_years}")`.  Try to explain this as clearly as possible for 
# your fellow students.  


# the $db.execute statement, will execute the SQL statement that we pass in the parantheses () and stores the result in an array. 



# If you're having trouble, find someone to pair on this explanation with you.

# What parts of your strategy worked? What problems did you face?

# Initially I had some trouble when I ran the command sqlite3 congress_poll_results.db, but then figured out what I was doing wrong.
# Also when I was debugging the code I had to run it multiple times before I got it right. I realized when you store the result of the 'db.execute'
# in a variable its a nested array.

# What questions did you have while coding? What resources did you find to help you answer them?

# Just had to execute the code multiple times before I understood how to fix it.

# What concepts are you having trouble with, or did you just figure something out? If so, what?

# Since this was the first time I was running code with db.execute statements it took some time. I had fun though with knowing how the Ruby code
#interacts with Database

# Did you learn any new skills or tricks?

# How to use Ruby code to interact with Database

# How confident are you with each of the Learning Competencies?

# Very confident 

# Which parts of the challenge did you enjoy?

# Coding and writing sql queries is always fun

# Which parts of the challenge did you find tedious?

# None